cupy/_core/_kernel.pyx

```diff
@@ -1584,21 +1613,18 @@ cdef class _Ops:
         return _Ops(tuple(ops_))
 
     cpdef _Op guess_routine(
-            self, str name, dict cache, list in_args, dtype, _Ops out_ops):
+            self, str name, dict cache, list in_args, tuple weaks, dtype,
+            _Ops out_ops):
         cdef _Ops ops_
         if dtype is None:
-            use_raw_value = _check_should_use_min_scalar(in_args)
-            if use_raw_value:
-                in_types = tuple([
-                    a.dtype.type if isinstance(a, _ndarray_base)
-                    else _min_scalar_type(a)
-                    for a in in_args])
-            else:
-                in_types = tuple([a.dtype.type for a in in_args])
-            op = cache.get(in_types, ())
+            assert all([isinstance(a, (_ndarray_base, numpy.generic))
+                        for a in in_args])
+            in_types = tuple([a.dtype.type for a in in_args])
+
+            op = cache.get((in_types, weaks), ())
```

Comment:
Makes sense, I suspect one could use `in_types[weak_pos] = int` or so also here, but I think using the `weak` tuple works well for starters.
(I might want to push for a cleanup after/with my other PR, but this is more urgent.)